### ISSUE TITLE
opencl apparently broken in latest nvidia-driver on Debian Sid

### ISSUE THREAD

**Issue by mir07** [darktable-org/darktable#13714]
This is tested and produces identical results on 4.2.0 and 4.2.1. Hope for some directions for a way to fix this annoyance: 

```
darktable --version
this is darktable 4.2.1
copyright (c) 2009-2023 johannes hanika
https://github.com/darktable-org/darktable/issues/new/choose

compile options:
  bit depth is 64 bit
  normal build
  SSE2 optimized codepath enabled
  OpenMP support enabled
  OpenCL support enabled
  Lua support enabled, API version 9.0.0
  Colord support enabled
  gPhoto2 support enabled
  GraphicsMagick support enabled
  ImageMagick support disabled
  libavif support enabled
  libheif support enabled
  libjxl support enabled
  OpenJPEG support enabled
  OpenEXR support enabled
  WebP support enabled

dpkg -s nvidia-driver-bin
Package: nvidia-driver-bin
Status: install ok installed
Priority: optional
Section: non-free/x11
Installed-Size: 350
Maintainer: Debian NVIDIA Maintainers <pkg-nvidia-devel@lists.alioth.debian.org>
Architecture: amd64
Source: nvidia-graphics-drivers
Version: 525.85.12-1
Provides: nvidia-driver-bin-525.85.12
Depends: nvidia-alternative (= 525.85.12-1), libc6 (>= 2.2.5), libnvidia-ml1 (>= 319) | libnvidia-ml.so.1 (>= 319)
Recommends: nvidia-driver
Conflicts: nvidia-driver-bin-525.85.12
Description: NVIDIA driver support binaries
 The NVIDIA binary driver provides optimized hardware acceleration of
 OpenGL/GLX/EGL/GLES applications via a direct-rendering X Server
 for graphics cards using NVIDIA chip sets.
 .
 This package contains supporting binaries for the driver.
Homepage: https://www.nvidia.com

dpkg -s nvidia-opencl-common
Package: nvidia-opencl-common
Status: install ok installed
Priority: optional
Section: non-free/libs
Installed-Size: 160
Maintainer: Debian NVIDIA Maintainers <pkg-nvidia-devel@lists.alioth.debian.org>
Architecture: amd64
Multi-Arch: foreign
Source: nvidia-graphics-drivers
Version: 525.85.12-1
Suggests: nvidia-opencl-icd | nvidia-tesla-opencl-icd | nvidia-tesla-470-opencl-icd
Conffiles:
 /etc/OpenCL/vendors/nvidia.icd 9d950bb477202d049cf31f84621a7f47
Description: NVIDIA OpenCL driver - common files
 OpenCL (Open Computing Language) is a multivendor open standard for
 general-purpose parallel programming of heterogeneous systems that include
 CPUs, GPUs and other processors.
 .
 This package provides the common files for the NVIDIA installable client
 driver (ICD) for OpenCL.
Homepage: https://www.nvidia.com

dpkg -s nvidia-opencl-icd:amd64
Package: nvidia-opencl-icd
Status: install ok installed
Priority: optional
Section: non-free/libs
Installed-Size: 22616
Maintainer: Debian NVIDIA Maintainers <pkg-nvidia-devel@lists.alioth.debian.org>
Architecture: amd64
Multi-Arch: same
Source: nvidia-graphics-drivers
Version: 525.85.12-1
Provides: opencl-icd
Depends: nvidia-opencl-common, ocl-icd-libopencl1 | nvidia-libopencl1 | libopencl1, nvidia-alternative (= 525.85.12-1), libcuda1 (= 525.85.12-1), libnvidia-compiler (= 525.85.12-1), libc6 (>= 2.34)
Enhances: libopencl1
Description: NVIDIA OpenCL installable client driver (ICD)
 OpenCL (Open Computing Language) is a multivendor open standard for
 general-purpose parallel programming of heterogeneous systems that include
 CPUs, GPUs and other processors.
 .
 This package provides the NVIDIA installable client driver (ICD) for OpenCL
 which supports NVIDIA GPUs.
Homepage: https://www.nvidia.com

darktable-cltest -d opencl -d verbose
[dt_get_sysresource_level] switched to 2 as `large'
  total mem:       32029MB
  mipmap cache:    4003MB
  available mem:   21895MB
  singlebuff:      500MB
  OpenCL tune mem: WANTED
  OpenCL pinned:   WANTED
[opencl_init] opencl related configuration options:
[opencl_init] opencl: ON
[opencl_init] opencl_scheduling_profile: 'default'
[opencl_init] opencl_library: 'default path'
[opencl_init] opencl_device_priority: 'NVIDIA GeForce GTX 1050 Ti,*/!NVIDIA GeForce GTX 1050 Ti,*/NVIDIA GeForce GTX 1050 Ti,*/NVIDIA GeForce GTX 1050 Ti,*'
[opencl_init] opencl_mandatory_timeout: 200
0.033757 [dt_dlopencl_init] found opencl runtime library 'libOpenCL'
[opencl_init] opencl library 'libOpenCL' found on your system and loaded
[opencl_init] found 3 platforms
[opencl_init] no devices found for Mesa (vendor) - Clover (name)
[opencl_init] no devices found for Mesa/X.org (vendor) - rusticl (name)
[opencl_init] found 1 device

[dt_opencl_device_init]
0.138019 [dt_opencl_write_device_config] writing data '0 250 0 16 16 128 0 0 0.023895 0.000' for 'cldevice_v4_nvidiageforcegtx1050ti'
0.138026 [dt_opencl_write_device_config] writing data '400' for 'cldevice_v4_nvidiageforcegtx1050ti_id0'
   DEVICE:                   0: 'NVIDIA GeForce GTX 1050 Ti'
   CANONICAL NAME:           nvidiageforcegtx1050ti
   PLATFORM NAME & VENDOR:   NVIDIA CUDA, NVIDIA Corporation
   DRIVER VERSION:           525.85.12
   DEVICE VERSION:           OpenCL 3.0 CUDA, SM_20 SUPPORT
   DEVICE_TYPE:              GPU
   GLOBAL MEM SIZE:          4031 MB
   MAX MEM ALLOC:            1008 MB
   MAX IMAGE SIZE:           16384 x 32768
   MAX WORK GROUP SIZE:      1024
   MAX WORK ITEM DIMENSIONS: 3
   MAX WORK ITEM SIZES:      [ 1024 1024 64 ]
   ASYNC PIXELPIPE:          NO
   PINNED MEMORY TRANSFER:   WANTED
   MEMORY TUNING:            WANTED
   FORCED HEADROOM:          400
   AVOID ATOMICS:            NO
   MICRO NAP:                250
   ROUNDUP WIDTH:            16
   ROUNDUP HEIGHT:           16
   CHECK EVENT HANDLES:      128
   PERFORMANCE:              4.737
   TILING ADVANTAGE:         0.000
   DEFAULT DEVICE:           NO
   KERNEL DIRECTORY:         /usr/share/darktable/kernels
   CL COMPILER OPTION:       -cl-fast-relaxed-math
0.193404 [dt_opencl_device_init] testing program `demosaic_ppg.cl' ..
0.193472 [opencl_fopen_stat] could not open file `/home/mir/.cache/darktable/cached_kernels_for_NVIDIAGeForceGTX1050Ti_5258512/demosaic_ppg.cl.bin'!
0.193480 [opencl_load_program] could not load cached binary program, trying to compile source
0.193490 [opencl_load_program] successfully loaded program from '/usr/share/darktable/kernels/demosaic_ppg.cl' MD5: 'ae37feb4062866f41669d99c8dc0b2dd'
0.193651 [opencl_build_program] could not build program: CL_BUILD_PROGRAM_FAILURE
0.193656 [opencl_build_program] BUILD STATUS: -1
0.193658 BUILD LOG:
0.193659 
0.193660 [dt_opencl_device_init] failed to compile program `demosaic_ppg.cl'!
0.193667 [dt_opencl_write_device_config] writing data '0 250 0 16 16 128 0 0 0.023895 0.000' for 'cldevice_v4_nvidiageforcegtx1050ti'
0.193672 [dt_opencl_write_device_config] writing data '400' for 'cldevice_v4_nvidiageforcegtx1050ti_id0'
[opencl_init] no suitable devices found.
[opencl_init] FINALLY: opencl is NOT AVAILABLE on this system.
[opencl_init] initial status of opencl enabled flag is OFF.
```

**Comment by TurboGit** [comment]
Yes, see my message here:

https://discuss.pixls.us/t/fixed-gnu-debian-nvidia-opencl-520-x-525-x/35321/11

**Comment by mir07** [comment]
Nice to hear. Looking forward to the needed fix to be released from the Debian team since working without OpenCL really demonstrate why using OpenCL is a blessing ;-)

Thank you for your response.

**Comment by TurboGit** [comment]
Hope you've seen that there is a simple workaround by setting a variable? I'm using this for now to at least have OpenCL because as you said it is quite important :)

**Comment by mir07** [comment]
> Hope you've seen that there is a simple workaround by setting a variable? I'm using this for now to at least have OpenCL because as you said it is quite important :)

Yes, you have saved my day, and for sure also my coming weekend ;-)
For others reading this is the fix:

- sudo apt install libnvidia-nvvm4
- cd /usr/lib/x86_64-linux-gnu
- sudo ln -s /usr/lib/x86_64-linux-gnu/nvidia/current/libnvidia-nvvm.so.4
- sudo ln -s /usr/lib/x86_64-linux-gnu/nvidia/current/libnvidia-nvvm.so.525.85.12

**Comment by TurboGit** [comment]
I prefer this:

`$ export LD_LIBRARY_PATH=/usr/lib/x86_64-linux-gnu/nvidia/current/`

As it doesn't changes install that will break further updates if not cleaned-up.

**Comment by mir07** [comment]
You are correct but to have it persisted add this to your .bashrc file:

# Fix for nvidia 5.20 and 5.25
export LD_LIBRARY_PATH=/usr/lib/x86_64-linux-gnu/nvidia/current/

And the do a re-login

**Comment by Nilvus** [comment]
As this is upstream bug, maybe we could closed it now?
